File: flighttracker/trips.py
"""Round-trip assembly: pair outbound legs (X->ALC) with return legs (ALC->X),
score each pairing, and mix in Luxair's whole-trip fares.

Shared by the web API, the alert evaluator and the chat context so that all
three agree on what "a trip" is and what it costs.
"""
from __future__ import annotations

import dataclasses
from datetime import date, datetime, timedelta

from . import db
from .config import Config
from .scoring import convenience_adjustment, day_adjustment, work_days_used


@dataclasses.dataclass
class TripFilter:
    min_nights: int = 1
    max_nights: int = 21
    airport: str = ""          # home airport (LUX/SCN/HHN); "" = any
    when: str = ""             # "", "weekend", "fri", "convenient", "weekday"
    same_only: bool = False    # return to the same airport you left from
    direct_only: bool = False
    max_price: float | None = None
    max_days_off: int | None = None   # work days the trip may consume


def matches_when(label: str, when: str) -> bool:
    if not when:
        return True
    if when == "weekend":
        return "weekend" in label
    if when == "fri":
        return "fri evening" in label
    if when == "convenient":
        return "weekend" in label or "fri evening" in label
    if when == "weekday":
        return "weekday" in label or "work hours" in label
    return True


def trip_key(combo: dict) -> str:
    """Stable identity for a pairing, used to remember alert hits."""
    return "|".join([
        combo["out"]["origin"], combo["out"]["departure"],
        combo["ret"]["destination"], combo["ret"]["departure"],
    ])


def build_trips(cfg: Config, f: TripFilter) -> tuple[list[dict], str | None]:
    """All pairings that pass the filter, cheapest-effective first, plus the
    timestamp of the most recent capture they were built from."""
    conn = db.connect()
    try:
        rows = db.latest_snapshot(conn)
        packages = db.latest_packages(conn)
    finally:
        conn.close()

    outbounds, returns, last_captured = [], [], None
    for r in rows:
        try:
            dep = datetime.fromisoformat(r["departure"])
        except ValueError:
            continue
        adj, label = convenience_adjustment(dep, cfg.scoring)
        leg = {
            "origin": r["origin"], "destination": r["destination"],
            "departure": r["departure"], "airline": r["airline"],
            "stops": r["stops"],
            "price": r["price"], "adjustment": adj, "label": label,
            "source": r["source"], "_dep": dep,
        }
        if last_captured is None or r["captured_at"] > last_captured:
            last_captured = r["captured_at"]
        if f.direct_only and r["stops"]:
            continue
        if r["destination"] == "ALC":
            if f.airport and r["origin"] != f.airport:
                continue
            if not matches_when(label, f.when):
                continue
            outbounds.append(leg)
        elif r["origin"] == "ALC":
            if f.airport and r["destination"] != f.airport:
                continue
            returns.append(leg)

    combos = []
    for out in outbounds:
        for ret in returns:
            nights = (ret["_dep"].date() - out["_dep"].date()).days
            if nights < f.min_nights or nights > f.max_nights:
                continue
            if nights == 0 and ret["_dep"] <= out["_dep"]:
                continue
            if f.same_only and out["origin"] != ret["destination"]:
                continue
            price = out["price"] + ret["price"]
            if f.max_price is not None and price > f.max_price:
                continue
            days_off = work_days_used(out["_dep"], ret["_dep"], cfg.scoring)
            if f.max_days_off is not None and days_off > f.max_days_off:
                continue
            combos.append({
                "out": {k: v for k, v in out.items() if k != "_dep"},
                "ret": {k: v for k, v in ret.items() if k != "_dep"},
                "nights": nights,
                "days_off": days_off,
                "same_airport": out["origin"] == ret["destination"],
                "price": round(price, 2),
                "adjustment": out["adjustment"] + ret["adjustment"],
                "effective": round(price + out["adjustment"] + ret["adjustment"], 2),
            })

    combos.extend(_package_combos(packages, cfg, f))
    combos.sort(key=lambda c: c["effective"])
    return combos, last_captured


def _package_combos(packages, cfg: Config, f: TripFilter) -> list[dict]:
    """Luxair round trips, shaped like the paired ones so the UI can mix them.

    Non-stop and quoted whole, but with no departure times, so they are scored
    on the day of week only (scoring.day_adjustment) and their days-off count
    is conservative (scoring.work_days_used with date_only=True)."""
    out = []
    for r in packages:
        if f.airport and r["origin"] != f.airport and r["destination"] != f.airport:
            continue
        if not (f.min_nights <= r["nights"] <= f.max_nights):
            continue
        if f.max_price is not None and r["price"] > f.max_price:
            continue
        try:
            out_day = date.fromisoformat(r["out_date"])
        except ValueError:
            continue
        ret_day = out_day + timedelta(days=r["nights"])
        out_adj, out_label = day_adjustment(out_day, cfg.scoring)
        ret_adj, ret_label = day_adjustment(ret_day, cfg.scoring)
        if not matches_when(out_label, f.when):
            continue
        out_dt = datetime.combine(out_day, datetime.min.time())
        ret_dt = datetime.combine(ret_day, datetime.min.time())
        days_off = work_days_used(out_dt, ret_dt, cfg.scoring, True, True)
        if f.max_days_off is not None and days_off > f.max_days_off:
            continue
        out.append({
            "out": {
                "origin": r["origin"], "destination": r["destination"],
                "departure": out_dt.isoformat(), "date_only": True,
                "airline": "Luxair", "stops": 0, "price": None,
                "adjustment": out_adj, "label": out_label, "source": "luxair",
            },
            "ret": {
                "origin": r["destination"], "destination": r["origin"],
                "departure": ret_dt.isoformat(), "date_only": True,
                "airline": "Luxair", "stops": 0, "price": None,
                "adjustment": ret_adj, "label": ret_label, "source": "luxair",
            },
            "nights": r["nights"],
            "days_off": days_off,
            "same_airport": True,
            "package": True,
            "price": r["price"],
            "adjustment": out_adj + ret_adj,
            "effective": round(r["price"] + out_adj + ret_adj, 2),
        })
    return out

File: src/lib/trips.ts
import type { Config } from "./config";
import {
  airportGround,
  arrivalAdjustment,
  convenienceAdjustment,
  dayAdjustment,
  daysOffCost,
  tripGround,
  workDaysUsed,
} from "./scoring";
import {
  addDays,
  canonical,
  dayNumber,
  formatWallClock,
  parseWallClock,
  serial,
  startOfDay,
  type WallClock,
} from "./wallclock";

/**
 * Round-trip assembly: pair outbound legs (X→ALC) with return legs (ALC→X),
 * score each pairing, and mix in Luxair's whole-trip fares. Port of
 * flighttracker/trips.py.
 *
 * Pure: it takes the latest snapshot as rows and returns trips. The database
 * query lives in src/db/queries.ts so that this can be run over a fixture —
 * which is how tests/equivalence.test.ts proves it agrees with the Python
 * copy that the fetcher's alert check still uses.
 *
 * Effective cost = ticket + convenience adjustment + ground (the drive to
 * the airport and parking) + holiday (the work days the trip eats beyond
 * the departure day). Each part is kept on the trip so the UI can spell the
 * ranking out.
 *
 * Iteration order is part of the contract. Trips are built outbound-major
 * in snapshot order, packages appended, then stably sorted by effective
 * cost. Python does the same, and stable sorts on both sides mean ties
 * resolve identically — so the top-N an alert records is the same top-N.
 */

export interface FareRow {
  origin: string;
  destination: string;
  /** Wall clock, "T" or space separated. */
  departure: string;
  arrival?: string | null;
  airline: string | null;
  stops: number;
  price: number;
  currency: string;
  source: string;
  /** ISO instant, comparable as text. */
  capturedAt: string;
}

export interface PackageRow {
  origin: string;
  destination: string;
  /** "YYYY-MM-DD" */
  outDate: string;
  nights: number;
  price: number;
  currency: string;
  source: string;
  capturedAt: string;
}

export type When = "" | "weekend" | "fri" | "convenient" | "weekday";

export interface TripFilter {
  minNights: number;
  maxNights: number;
  /** Home airport (LUX/SCN/HHN); "" = any. */
  airport: string;
  when: When;
  /** Return to the same airport you left from. */
  sameOnly: boolean;
  directOnly: boolean;
  maxPrice: number | null;
  /** Work days the trip may consume. */
  maxDaysOff: number | null;
}

export const DEFAULT_FILTER: TripFilter = {
  minNights: 1,
  maxNights: 21,
  airport: "",
  when: "",
  sameOnly: false,
  directOnly: false,
  maxPrice: null,
  maxDaysOff: null,
};

export interface Leg {
  origin: string;
  destination: string;
  /** Canonical "YYYY-MM-DDTHH:MM:SS". */
  departure: string;
  arrival: string | null;
  airline: string | null;
  stops: number;
  /** null for a package leg: the price belongs to the whole trip. */
  price: number | null;
  adjustment: number;
  label: string;
  source: string;
  /** Luxair: no departure time is known. */
  dateOnly?: boolean;
}

export interface Trip {
  out: Leg;
  ret: Leg;
  nights: number;
  daysOff: number;
  sameAirport: boolean;
  isPackage: boolean;
  price: number;
  adjustment: number;
  /** Driving and parking, euros. */
  ground: number;
  groundLabel: string;
  /** Holiday charged, euros. */
  holiday: number;
  holidayLabel: string;
  effective: number;
}

export function matchesWhen(label: string, when: When): boolean {
  if (!when) return true;
  if (when === "weekend") return label.includes("weekend");
  if (when === "fri") return label.includes("fri evening");
  if (when === "convenient")
    return label.includes("weekend") || label.includes("fri evening");
  if (when === "weekday")
    return label.includes("weekday") || label.includes("work hours");
  return true;
}

/**
 * Stable identity for a pairing, used to remember alert hits. Must be byte
 * for byte what Python's trip_key produces: the fetcher writes these into
 * alert_hits and the app reads them back.
 */
export function tripKey(trip: Trip): string {
  return [
    trip.out.origin,
    trip.out.departure,
    trip.ret.destination,
    trip.ret.departure,
  ].join("|");
}

/** Python's round(x, 2) on values that never land on an exact half. */
function round2(x: number): number {
  return Math.round(x * 100) / 100;
}

function parseOrNull(text: string | null | undefined): WallClock | null {
  if (!text) return null;
  try {
    return parseWallClock(text);
  } catch {
    return null;
  }
}

interface ScoredLeg extends Leg {
  dep: WallClock;
}

/**
 * Score one snapshot row as a leg: convenience of the departure, plus the
 * late-arrival penalty when it lands back home at night. Shared with the
 * one-way list and the calendar, which score legs the same way.
 */
export function scoreLeg(r: FareRow, cfg: Config): ScoredLeg | null {
  let dep: WallClock;
  try {
    dep = parseWallClock(r.departure);
  } catch {
    return null;
  }
  let [adjustment, label] = convenienceAdjustment(dep, cfg.scoring);
  const [arrAdj, arrLabel] = arrivalAdjustment(
    parseOrNull(r.arrival),
    r.destination !== "ALC",
    cfg.scoring,
  );
  if (arrLabel) {
    adjustment += arrAdj;
    label = `${label}, ${arrLabel}`;
  }
  return {
    origin: r.origin,
    destination: r.destination,
    departure: canonical(r.departure),
    arrival: r.arrival ? canonical(r.arrival) : null,
    airline: r.airline,
    stops: r.stops,
    price: r.price,
    adjustment,
    label,
    source: r.source,
    dep,
  };
}

/**
 * All pairings that pass the filter, cheapest-effective first, plus the
 * timestamp of the most recent capture they were built from.
 */
export function buildTrips(
  rows: FareRow[],
  packages: PackageRow[],
  cfg: Config,
  f: TripFilter,
): { trips: Trip[]; lastCaptured: string | null } {
  const outbounds: ScoredLeg[] = [];
  const returns: ScoredLeg[] = [];
  let lastCaptured: string | null = null;

  for (const r of rows) {
    const leg = scoreLeg(r, cfg);
    if (!leg) continue;
    // Before the filters, as in Python: the snapshot age is a property of
    // the data, not of what the user asked to see.
    if (lastCaptured === null || r.capturedAt > lastCaptured) {
      lastCaptured = r.capturedAt;
    }
    if (f.directOnly && r.stops) continue;
    if (r.destination === "ALC") {
      if (f.airport && r.origin !== f.airport) continue;
      if (!matchesWhen(leg.label, f.when)) continue;
      outbounds.push(leg);
    } else if (r.origin === "ALC") {
      if (f.airport && r.destination !== f.airport) continue;
      returns.push(leg);
    }
  }

  const trips: Trip[] = [];
  for (const out of outbounds) {
    for (const ret of returns) {
      const nights = dayNumber(ret.dep) - dayNumber(out.dep);
      if (nights < f.minNights || nights > f.maxNights) continue;
      if (nights === 0 && serial(ret.dep) <= serial(out.dep)) continue;
      if (f.sameOnly && out.origin !== ret.destination) continue;
      const price = (out.price as number) + (ret.price as number);
      if (f.maxPrice !== null && price > f.maxPrice) continue;
      const daysOff = workDaysUsed(out.dep, ret.dep, cfg.scoring);
      if (f.maxDaysOff !== null && daysOff > f.maxDaysOff) continue;
      const adjustment = out.adjustment + ret.adjustment;
      const [ground, groundLabel] = tripGround(out.origin, ret.destination, nights, cfg.travel);
      const [holiday, holidayLabel] = daysOffCost(daysOff, out.dep, cfg.scoring);
      trips.push({
        out: stripDep(out),
        ret: stripDep(ret),
        nights,
        daysOff,
        sameAirport: out.origin === ret.destination,
        isPackage: false,
        price: round2(price),
        adjustment,
        ground,
        groundLabel,
        holiday,
        holidayLabel,
        effective: round2(price + adjustment + ground + holiday),
      });
    }
  }

  trips.push(...packageTrips(packages, cfg, f));
  // Array.prototype.sort is stable (ES2019), as Python's list.sort is.
  trips.sort((a, b) => a.effective - b.effective);
  return { trips, lastCaptured };
}

function stripDep(leg: ScoredLeg): Leg {
  return {
    origin: leg.origin,
    destination: leg.destination,
    departure: leg.departure,
    arrival: leg.arrival,
    airline: leg.airline,
    stops: leg.stops,
    price: leg.price,
    adjustment: leg.adjustment,
    label: leg.label,
    source: leg.source,
  };
}

/**
 * Luxair round trips, shaped like the paired ones so the UI can mix them.
 *
 * Non-stop and quoted whole, but with no departure times, so they are
 * scored on the day of week only (dayAdjustment) and their days-off count
 * is conservative (workDaysUsed with dateOnly).
 */
function packageTrips(
  packages: PackageRow[],
  cfg: Config,
  f: TripFilter,
): Trip[] {
  const out: Trip[] = [];
  for (const r of packages) {
    if (f.airport && r.origin !== f.airport && r.destination !== f.airport)
      continue;
    if (!(f.minNights <= r.nights && r.nights <= f.maxNights)) continue;
    if (f.maxPrice !== null && r.price > f.maxPrice) continue;
    let outDay: WallClock;
    try {
      outDay = startOfDay(parseWallClock(r.outDate));
    } catch {
      continue;
    }
    const retDay = addDays(outDay, r.nights);
    const [outAdj, outLabel] = dayAdjustment(outDay, cfg.scoring);
    const [retAdj, retLabel] = dayAdjustment(retDay, cfg.scoring);
    if (!matchesWhen(outLabel, f.when)) continue;
    const daysOff = workDaysUsed(outDay, retDay, cfg.scoring, true, true);
    if (f.maxDaysOff !== null && daysOff > f.maxDaysOff) continue;
    const home = r.origin === "ALC" ? r.destination : r.origin;
    const [ground, groundLabel] = airportGround(home, cfg.travel, r.nights);
    const [holiday, holidayLabel] = daysOffCost(daysOff, outDay, cfg.scoring, true);
    out.push({
      out: {
        origin: r.origin,
        destination: r.destination,
        departure: formatWallClock(outDay),
        arrival: null,
        dateOnly: true,
        airline: "Luxair",
        stops: 0,
        price: null,
        adjustment: outAdj,
        label: outLabel,
        source: "luxair",
      },
      ret: {
        origin: r.destination,
        destination: r.origin,
        departure: formatWallClock(retDay),
        arrival: null,
        dateOnly: true,
        airline: "Luxair",
        stops: 0,
        price: null,
        adjustment: retAdj,
        label: retLabel,
        source: "luxair",
      },
      nights: r.nights,
      daysOff,
      sameAirport: true,
      isPackage: true,
      price: r.price,
      adjustment: outAdj + retAdj,
      ground,
      groundLabel,
      holiday,
      holidayLabel,
      effective: round2(r.price + outAdj + retAdj + ground + holiday),
    });
  }
  return out;
}
